### DISCUSSION TITLE
Handle return "validation" errors

### DISCUSSION THREAD

**Discussion by gdaybrice** [next-safe-action/next-safe-action#256] [Q&A] [answered]
Is there a method similar to `handleReturnedServerError` for validation errors? I would like to have a single method to handle both server and validation errors and check for the error type. I'm using zod in other parts of my backend and need to handle that logic in both places otherwise.

I've seen the `handleValidationErrorsShape` method but I want to set this globally, not on every schema.

**Comment by MrOxMasTer** [comment]
In fact, I have been suggesting to the author for a very long time to make a validation error handling function, but unfortunately the author of the library ignores me.

https://github.com/TheEdoRan/next-safe-action/discussions/206

I don't remember what I wrote then, but there really isn't enough function like `handleServerError`

You can call it something like this: `handleValidationError`

You can also add the second arguments of `utils` as in `handleServerError`, so that there would be something to return. And the first argument above is this `ve`, like 'handleValidationErrorShape'.

This feature can be added both for individual schemes and for the general client

**Comment by TheEdoRan** [comment, marked as answer]
Hi, the server error shape is a global setting because all the actions share the same returned server error type. However, validation errors are determined by the schema you pass to the action you're defining, that's why it wouldn't make any sense to set an option at the global client level, other than a generic `"flattened" | "formatted"` default shape. The client is used to define all the actions of your application, there's just no way to know the specific schemas (and so, validation errors types) of all your actions at the top level of the hierarchy, it's impossible. Hope this clears things up!

**Reply by MrOxMasTer** [reply]
> Hi, the server error shape is a global setting because all the actions share the same returned server error type. However, validation errors are determined by the schema you pass to the action you're defining, that's why it wouldn't make any sense to set an option at the global client level, other than a generic `"flattened" | "formatted"` default shape. The client is used to define all the actions of your application, there's just no way to know the specific schemas (and so, validation errors types) of all your actions at the top level of the hierarchy, it's impossible. Hope this clears things up!

Well there is a function to determine for each circuit the type of errors. You can try to make a template, for example `flattenValidationErrors(ve).fieldErrors`.

![image](https://github.com/user-attachments/assets/892621f9-362e-4fab-be25-620171dba3d2)

But if it's my question, why can't we add additional fields, when validation errors occur.  Or a function to handle validation errors.
In case of validation errors, for example, sending a request or additional fields (utils).

**Reply by TheEdoRan** [reply]
> Well there is a function to determine for each circuit the type of errors. You can try to make a template, for example `flattenValidationErrors(ve).fieldErrors`.

As I said, that's specific to the action schema, it's not at the global level. There's no way to know which fields each action will have, it's also impossible to infer types.

> But if it's my question, why can't we add additional fields, when validation errors occur. Or a function to handle validation errors.
In case of validation errors, for example, sending a request or additional fields.

I don't understand what you're trying to say.

**Reply by MrOxMasTer** [reply]
> I don't understand what you're trying to say.

My point is that the validation is very axe-headed. And if you can customize it for each scheme, it would be nice to have a validation error handling function for each scheme. 

`variant 1`
```ts
const checkUpUserAction = actionClient
  .schema(formDataCheckUpUserSchema, {
     handleValidationError: (utils) => {
          const { clientInput, bindArgsClientInputs, metadata, ctx } = utils;

         // Anything
        // Returns the data entered by the user
        // Sending context (may get lost due to validation issues)

           return {
              clientInput
           }
     }
  })
  .stateAction(
    async ({ parsedInput: { email } }) => {
      ...
      return user;
    },
    {
      onError: ({ clientInput }) => {
        ...
      },
      onSuccess: ({ data }) => {
        ...
      },
    },
  );
```

`variant 2`
```ts
const checkUpUserAction = actionClient
  .schema(formDataCheckUpUserSchema)
  .stateAction(
    async ({ parsedInput: { email } }) => {
      ...
      return user;
    },
    {
      onError: ({ clientInput }) => {
        ...

        return clientInput;
      },
      onSuccess: ({ data }) => {
        ...
      },
    },
  );
```

**Reply by MrOxMasTer** [reply]
> I don't understand what you're trying to say.

That's why I suggested the options here. The 2 best ones are either to make a separate function like `handleValidationError` in the schema itself, or to let users return something from `action callbacks`, as you have in your documentation that: 
![image](https://github.com/user-attachments/assets/a7ea9b91-a7c3-4fb4-8ea1-7d20394b53e8)

**Reply by MrOxMasTer** [reply]
> I don't understand what you're trying to say.

Okay, so I thought about it and realized that I can perform actions in case of a validation error, but the problem is that I don't have any contact with the client. If there is a validation error, I cannot return anything additional to the client, except for the validation error itself made by the library, and this is a problem if I want to return, for example, what the user sent or the context to the client in `state`

I will put this in a separate discussion.

**Reply by TheEdoRan** [reply]
@MrOxMasTer Ok, it's clearer now. I think I can pass an `utils` object with additional data (such as `clientInput`, `metadata`, `ctx`, etc.) to `handleValidationErrorsShape` function, found in `schema` method (your "variant 1" code snippet).

**Reply by MrOxMasTer** [reply]
> @MrOxMasTer Ok, it's clearer now. I think I can pass an `utils` object with additional data (such as `clientInput`, `metadata`, `ctx`, etc.) to `handleValidationErrorsShape` function, found in `schema` method (your "variant 1" code snippet).

Please read this. That's the problem I'm talking about:
https://github.com/TheEdoRan/next-safe-action/discussions/260

It would be great to make `useStateAction`, but as I understand it, without changing `react' this will not work

**Reply by MrOxMasTer** [reply]
> @MrOxMasTer Ok, it's clearer now. I think I can pass an `utils` object with additional data (such as `clientInput`, `metadata`, `ctx`, etc.) to `handleValidationErrorsShape` function, found in `schema` method (your "variant 1" code snippet).

so that he supports progressive improvement*

**Reply by TheEdoRan** [reply]
Starting from [v7.9.2](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.9.2), `utils` object is passed to both `handleValidationErrorsShape` and `handleBindArgsValidationErrorsShape` functions as the second argument. (#263)
